lib/assembler.cc

```diff
@@ -324,78 +329,190 @@ const
     }
 }
 
+/***************************************
+ * Junction-counting assembler
+ ***************************************/
 
+JunctionCountAssembler::JunctionCountAssembler(Hashgraph * ht) :
+    graph(ht), _ksize(ht->ksize()), traverser(ht), linear_asm(ht)
+{
+    std::vector<uint64_t> table_sizes = graph->get_tablesizes();
+    junctions = new CountingHash(_ksize, table_sizes);
+}
 
-/*
-template <bool direction>
-void SimpleLabeledAssembler::_assemble_directed(NonLoopingAT<direction>&
-        start_cursor,
-        StringVector& paths)
-const
+
+JunctionCountAssembler::~JunctionCountAssembler()
+{
+    delete this->junctions;
+}
+
+uint16_t JunctionCountAssembler::consume(std::string sequence)
 {
+    // first we need to put the sequence in the graph
+    graph->consume_string(sequence);
+    // now we find its high degree nodes and count the
+    // branch junctions
+    KmerIterator kmers(sequence.c_str(), _ksize);
+    Kmer kmer = kmers.next();
+    if (kmers.done()) {
+        return 0;
+    }
+    Kmer next_kmer = kmers.next();
+    if (kmers.done()) {
+        return 0;
+    }
+    uint16_t d = this->traverser.degree(kmer);
+    uint16_t next_d = this->traverser.degree(next_kmer);
+    uint16_t n_junctions = 0;
+
+    while(!kmers.done()) {
+        if (d > 2 || next_d > 2) {
+            count_junction(kmer, next_kmer);
+            n_junctions++;
 #if DEBUG_ASSEMBLY
-    std::cout << "## assemble_labeled_directed_" << direction << " [start] at " <<
-        start_cursor.cursor.repr(_ksize) << std::endl;
+            std::cout << "Junction: " << kmer.repr(_ksize) << ", " << next_kmer.repr(_ksize) << std::endl;
+            std::cout << "Junction Count: " << get_junction_count(kmer, next_kmer) << std::endl;
 #endif
-    std::string root_contig = linear_asm->_assemble_directed<direction>
-                              (start_cursor);
+        }
+        kmer = next_kmer;
+        d = next_d;
+        next_kmer = kmers.next();
+        next_d = this->traverser.degree(next_kmer);
+    }
 
-    if (start_cursor.cursor_degree() > 1) {               // hit a HDN
+    return n_junctions / 2; 
+}
+
+void JunctionCountAssembler::count_junction(Kmer kmer_a, Kmer kmer_b)
+{
+    junctions->count(kmer_a.kmer_u ^ kmer_b.kmer_u);
+}
+
+BoundedCounterType JunctionCountAssembler::get_junction_count(Kmer kmer_a, Kmer kmer_b)
+const
+{
+    return junctions->get_count(kmer_a.kmer_u ^ kmer_b.kmer_u);
+}
+
+// Starting from the given seed k-mer, assemble all maximal linear paths in
+// both directions, using labels to skip over tricky bits.
```

Comment:
This comment is also wrong - it's using junction counters to skip over tricky bits :)